gapis/api/vulkan/api/coherent_memory.api

```diff
@@ -56,58 +56,170 @@ sub void readCoherentMemory(ref!DeviceMemoryObject memory, VkDeviceSize readOffs
   mem := buffer.Memory
   if mem != null {
-    if as!u64(readSize) == 0xFFFFFFFFFFFFFFFF {
-      readCoherentMemory(mem, buffer.MemoryOffset + readOffset, buffer.Info.Size - readOffset)
-    } else {
-      readCoherentMemory(mem, buffer.MemoryOffset + readOffset, readSize)
+    readBegin := buffer.MemoryOffset + readOffset
+    actualReadSize := switch as!u64(readSize) {
+      case 0xFFFFFFFFFFFFFFFF:
+        buffer.Info.Size - readOffset
+      default:
+        readSize
+    }
+    readEnd := readBegin + actualReadSize
+    readCoherentMemory(mem, readBegin, actualReadSize)
+    if len(mem.Data) != 0 {
+      read(mem.Data[readBegin : readEnd])
     }
   }
 }
 
-sub void readCoherentMemoryInBufferBindings(map!(u32, ref!VkDescriptorBufferInfo) bufferBindings) {
-  for _, _, v in bufferBindings {
+sub void readMemoryInBufferBindings(map!(u32, ref!VkDescriptorBufferInfo) bufferBindings, map!(u32, VkDeviceSize) bufferBindingOffsets) {
+  for k, _, v in bufferBindings {
     if v.Buffer != as!VkBuffer(0) {
       if (v.Buffer in Buffers) {
-        readCoherentMemoryInBuffer(Buffers[v.Buffer], v.Offset, v.Range)
+        offset := switch as!u32(k) in bufferBindingOffsets {
+          case true:
+            bufferBindingOffsets[as!u32(k)]
+          case false:
+            v.Offset
+        }
+        readMemoryInBuffer(Buffers[v.Buffer], offset, v.Range)
       }
     }
   }
 }
 
-sub void readCoherentMemoryBufferViewBindings(map!(u32, VkBufferView) bufferViews) {
+sub void readMemoryInBufferViewBindings(map!(u32, VkBufferView) bufferViews) {
   for _, _, v in bufferViews {
     if v in BufferViews {
       view := BufferViews[v]
       if (view.Buffer.VulkanHandle in Buffers) {
-        readCoherentMemoryInBuffer(view.Buffer, view.Offset, view.Range)
+        readMemoryInBuffer(view.Buffer, view.Offset, view.Range)
       }
     }
   }
 }
 
-sub void readCoherentMemoryInDescriptorSet(ref!DescriptorSetObject descriptor_set) {
-  for _, _, v in descriptor_set.Bindings {
+sub void readMemoryInImageBindings(map!(u32, ref!VkDescriptorImageInfo) imageBindings) {
+  for _, _, v in imageBindings {
+    _ = Samplers[v.Sampler]
+    if v.ImageView != as!VkImageView(0) {
+      imageViewObj := ImageViews[v.ImageView]
+      imageObj := imageViewObj.Image
+      rng := imageViewObj.SubresourceRange
+      readImageSubresource(imageObj, rng, v.ImageLayout)
+    }
+  }
+}
+
+sub void readMemoryInDescriptorSet(ref!DescriptorSetObject descriptor_set, map!(u32, map!(u32, VkDeviceSize)) bufferBindingOffsets) {
+  for j, _, v in descriptor_set.Bindings {
     switch v.BindingType {
       case
         VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,
         VK_DESCRIPTOR_TYPE_STORAGE_BUFFER,
         VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC,
         VK_DESCRIPTOR_TYPE_STORAGE_BUFFER_DYNAMIC:
-          readCoherentMemoryInBufferBindings(v.BufferBinding)
+          readMemoryInBufferBindings(v.BufferBinding, bufferBindingOffsets[as!u32(j)])
```

Comment:
Looks like we want `j` to be the binding number here, but actually `j` is index. Probably we should use `for _, j, v in descriptor_set.Bindings {`?